### ISSUE TITLE
Keybindings dont work and do not show up in folke/which-key.nvim

### ISSUE THREAD

**Issue by matu3ba** [lewis6991/gitsigns.nvim#498]
### Description

my setup uses packer with
```lua
  use { 'lewis6991/gitsigns.nvim', config = function() require('gitsigns').setup() end }
```

I did also try to use the config with `  on_attach = function(bufnr)` etc, but that made no difference.

### Neovim version

NVIM v0.7.0-dev+1360-ga18c9ba2d

### Operating system and version

Linux pc 5.17.1-arch1-1

### Expected behavior

1. ] shows a possible completion with c for the possible jump to next hunk
2. <leader>hb shows git blame. (it works with `:Gitsigns blame_line` in the cmd window.)

### Actual behavior

1. ] does not should show a possible completion with c for the possible jump to next hunk
2. <leader>hb does not should show git blame. (it works with `:Gitsigns blame_line` in the cmd window.)

### Minimal config

```Lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
  -- ADD OTHER PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, -- You must add this to enable debug messages
  -- ADD GITSIGNS CONFIG THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
  require('gitsigns').setup{
    on_attach = function(bufnr)
      local gs = package.loaded.gitsigns

      local function map(mode, l, r, opts)
        opts = opts or {}
        opts.buffer = bufnr
        vim.keymap.set(mode, l, r, opts)
      end

      -- Navigation
      map('n', ']c', function()
        if vim.wo.diff then return ']c' end
        vim.schedule(function() gs.next_hunk() end)
        return '<Ignore>'
      end, {expr=true})

      map('n', '[c', function()
        if vim.wo.diff then return '[c' end
        vim.schedule(function() gs.prev_hunk() end)
        return '<Ignore>'
      end, {expr=true})

      -- Actions
      map({'n', 'v'}, '<Space>hs', ':Gitsigns stage_hunk<CR>')
      map({'n', 'v'}, '<Space>hr', ':Gitsigns reset_hunk<CR>')
      map('n', '<Space>hS', gs.stage_buffer)
      map('n', '<Space>hu', gs.undo_stage_hunk)
      map('n', '<Space>hR', gs.reset_buffer)
      map('n', '<Space>hp', gs.preview_hunk)
      map('n', '<Space>hb', function() gs.blame_line{full=true} end)
      map('n', '<Space>tb', gs.toggle_current_line_blame)
      map('n', '<Space>hd', gs.diffthis)
      map('n', '<Space>hD', function() gs.diffthis('~') end)
      map('n', '<Space>td', gs.toggle_deleted)

      -- Text object
      map({'o', 'x'}, 'ih', ':<C-U>Gitsigns select_hunk<CR>')
    end
  }
}
```

### Steps to reproduce

1. `nvim -nu minimal.lua`
2. ...

### Gitsigns debug messages

The minimal config with running `nvim -nu minimal.lua` gives me this error:
```txt
No message shown in either local folder.

/tmp. $HOME/.config/nvim shows only the undodir. 

$HOME/.local/share/nvim only lsp, plenary and packer stuff.
```

**Comment by lewis6991** [comment]
Are you sure this isn't a bug with which-key?

**Comment by lewis6991** [comment]
There isn't a debug file. The messages are maintained within Neovim.

**Comment by lewis6991** [comment]
And are you sure the keymaps are even defined?

**Comment by matu3ba** [comment]
> And are you sure the keymaps are even defined?

Well, I used the instructions from the README as now added above.
But for some reason it appears that the `on_attach` does not work. How can I debug this problem?
I dont use `on_attach` in nvim_lspconfig.

**Comment by lewis6991** [comment]
You don't get any output from `:Gitsigns debug_messages`?

**Comment by matu3ba** [comment]
examples:

<details>

```txt
sequence
1. ]c with a change does not work
2. <Space>tb does not show git log of line
3. <Space>hb does not show git log of line
4. :Gitsigns blame_line works

run_job: git --no-pager --version
fn: Deriving GitSignsAdd from DiffAdd
fn: Deriving GitSignsChange from DiffChange
fn: Deriving GitSignsDelete from DiffDelete
fn: Deriving GitSignsAddNr from GitSignsAdd
fn: Deriving GitSignsChangeNr from GitSignsChange
fn: Deriving GitSignsDeleteNr from GitSignsDelete
fn: Deriving GitSignsAddLn from DiffAdd
fn: Deriving GitSignsChangeLn from DiffChange
fn: Deriving GitSignsDeleteLn from DiffDelete
fn: Deriving GitSignsCurrentLineBlame from NonText
fn: Deriving GitSignsAddInline from TermCursor
fn: Deriving GitSignsDeleteInline from TermCursor
fn: Deriving GitSignsChangeInline from TermCursor
fn: Deriving GitSignsAddLnInline from GitSignsAddInline
fn: Deriving GitSignsChangeLnInline from GitSignsChangeInline
fn: Deriving GitSignsDeleteLnInline from GitSignsDeleteInline
fn: Deriving GitSignsAddLnVirtLn from GitSignsAddLn
fn: Deriving GitSignsChangeVirtLn from GitSignsChangeLn
fn: Deriving GitSignsDeleteVirtLn from GitSignsDeleteLn
fn: Deriving GitSignsAddLnVirtLnInLine from GitSignsAddLnInline
fn: Deriving GitSignsChangeVirtLnInLine from GitSignsChangeLnInline
fn: Deriving GitSignsDeleteVirtLnInLine from GitSignsDeleteLnInline
fn: Highlight GitSignsAdd is already defined
fn: Highlight GitSignsChange is already defined
fn: Highlight GitSignsDelete is already defined
fn: Highlight GitSignsAddNr is already defined
fn: Highlight GitSignsChangeNr is already defined
fn: Highlight GitSignsDeleteNr is already defined
fn: Highlight GitSignsAddLn is already defined
fn: Highlight GitSignsChangeLn is already defined
fn: Highlight GitSignsDeleteLn is already defined
fn: Highlight GitSignsCurrentLineBlame is already defined
fn: Highlight GitSignsAddInline is already defined
fn: Highlight GitSignsDeleteInline is already defined
fn: Highlight GitSignsChangeInline is already defined
fn: Highlight GitSignsAddLnInline is already defined
fn: Highlight GitSignsChangeLnInline is already defined
fn: Highlight GitSignsDeleteLnInline is already defined
fn: Highlight GitSignsAddLnVirtLn is already defined
fn: Highlight GitSignsChangeVirtLn is already defined
fn: Highlight GitSignsDeleteVirtLn is already defined
fn: Highlight GitSignsAddLnVirtLnInLine is already defined
fn: Highlight GitSignsChangeVirtLnInLine is already defined
fn: Highlight GitSignsDeleteVirtLnInLine is already defined
attach(1): Attaching (trigger=setup)
run_job: git --no-pager config user.name
attach(1): Attach in progress
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --git-dir=/home/user/dev/git/zig/chepa/.git -c core.quotepath=off ls-files
 --stage --others --exclude-standard --eol /home/user/dev/git/zig/chepa/src/main.zig
watch_gitdir(1): Watching git dir
run_job: git --no-pager --git-dir=/home/user/dev/git/zig/chepa/.git show :0:src/main.zig
update(1): updates: 1, jobs: 7
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --git-dir=/home/user/dev/git/zig/chepa/.git blame --contents - -L 9,+1 --l
ine-porcelain /home/user/dev/git/zig/chepa/src/main.zig


sequence:
only running ]c

run_job: git --no-pager --version
fn: Deriving GitSignsAdd from DiffAdd
fn: Deriving GitSignsChange from DiffChange
fn: Deriving GitSignsDelete from DiffDelete
fn: Deriving GitSignsAddNr from GitSignsAdd
fn: Deriving GitSignsChangeNr from GitSignsChange
fn: Deriving GitSignsDeleteNr from GitSignsDelete
fn: Deriving GitSignsAddLn from DiffAdd
fn: Deriving GitSignsChangeLn from DiffChange
fn: Deriving GitSignsDeleteLn from DiffDelete
fn: Deriving GitSignsCurrentLineBlame from NonText
fn: Deriving GitSignsAddInline from TermCursor
fn: Deriving GitSignsDeleteInline from TermCursor
fn: Deriving GitSignsChangeInline from TermCursor
fn: Deriving GitSignsAddLnInline from GitSignsAddInline
fn: Deriving GitSignsChangeLnInline from GitSignsChangeInline
fn: Deriving GitSignsDeleteLnInline from GitSignsDeleteInline
fn: Deriving GitSignsAddLnVirtLn from GitSignsAddLn
fn: Deriving GitSignsChangeVirtLn from GitSignsChangeLn
fn: Deriving GitSignsDeleteVirtLn from GitSignsDeleteLn
fn: Deriving GitSignsAddLnVirtLnInLine from GitSignsAddLnInline
fn: Deriving GitSignsChangeVirtLnInLine from GitSignsChangeLnInline
fn: Deriving GitSignsDeleteVirtLnInLine from GitSignsDeleteLnInline
fn: Highlight GitSignsAdd is already defined
fn: Highlight GitSignsChange is already defined
fn: Highlight GitSignsDelete is already defined
fn: Highlight GitSignsAddNr is already defined
fn: Highlight GitSignsChangeNr is already defined
fn: Highlight GitSignsDeleteNr is already defined
fn: Highlight GitSignsAddLn is already defined
fn: Highlight GitSignsChangeLn is already defined
fn: Highlight GitSignsDeleteLn is already defined
fn: Highlight GitSignsCurrentLineBlame is already defined
fn: Highlight GitSignsAddInline is already defined
fn: Highlight GitSignsDeleteInline is already defined
fn: Highlight GitSignsChangeInline is already defined
fn: Highlight GitSignsAddLnInline is already defined
fn: Highlight GitSignsChangeLnInline is already defined
fn: Highlight GitSignsDeleteLnInline is already defined
fn: Highlight GitSignsAddLnVirtLn is already defined
fn: Highlight GitSignsChangeVirtLn is already defined
fn: Highlight GitSignsDeleteVirtLn is already defined
fn: Highlight GitSignsAddLnVirtLnInLine is already defined
fn: Highlight GitSignsChangeVirtLnInLine is already defined
fn: Highlight GitSignsDeleteVirtLnInLine is already defined
attach(1): Attaching (trigger=setup)
run_job: git --no-pager config user.name
attach(1): Attach in progress
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
run_job: git --no-pager --git-dir=/home/user/dev/git/zig/chepa/.git -c core.quotepath=off ls-files
 --stage --others --exclude-standard --eol /home/user/dev/git/zig/chepa/src/main.zig
watch_gitdir(1): Watching git dir
run_job: git --no-pager --git-dir=/home/user/dev/git/zig/chepa/.git show :0:src/main.zig
update(1): updates: 1, jobs: 7
run_job: git --no-pager rev-parse --show-toplevel --absolute-git-dir --abbrev-ref HEAD
```

</details>

**Comment by matu3ba** [comment]
The problem is that `on_attach` does not get run at all for some reasaon, which I verified by using `print("inside on_attach")` and `print("outside on_attach")`.
This section gives kind of a hint:
```txt
lspconfig users treat the lspconfig config object as a form of project management, using on_attach to set up keybindings, autocommands, and even buffer options. This becomes problematic when you have two language servers fighting with the same on_attach. Many users want to conditionally dispatch certain functions to one server or the other, they want a single unified set of settings for a given project’s buffers.

Basically lspconfig will be merged into core, and the remainder will be split across two separate repos
```

I do have Zig lsp running in the background, which might somehow break things?!

**Comment by matu3ba** [comment]
my workaround is using (not checked fully for correctness):
```lua
---- gitsigns ---- in file git operations (:Gitsigns debug_messages)
-- mappings are workaround of https://github.com/lewis6991/gitsigns.nvim/issues/498
map('v', ']c', '<cmd>Gitsigns next_hunk<CR>', opts)
map('v', '[c', '<cmd>Gitsigns prev_hunk<CR>', opts)
map('n', '<leader>hs', '<cmd>Gitsigns stage_hunk<CR>', opts)
map('v', '<leader>hs', '<cmd>Gitsigns stage_hunk<CR>', opts)
map('n', '<leader>hr', '<cmd>Gitsigns reset_hunk<CR>', opts)
map('v', '<leader>hr', '<cmd>Gitsigns reset_hunk<CR>', opts)
map('n', '<leader>hS', '<cmd>Gitsigns stage_buffer<CR>', opts)
map('n', '<leader>hu', '<cmd>Gitsigns undo_stage_hunk<CR>', opts)
map('n', '<leader>hR', '<cmd>Gitsigns reset_buffer<CR>', opts)
map('n', '<leader>hp', '<cmd>Gitsigns preview_hunk<CR>', opts)
map('n', '<leader>hb', '<cmd>Gitsigns blame_line<CR>', opts)
map('n', '<leader>tb', '<cmd>Gitsigns toggle_current_line_blame<CR>', opts)
map('n', '<leader>hd', '<cmd>Gitsigns diffthis<CR>', opts)
map('n', '<leader>hD', '<cmd>Gitsigns diffthis "~"<CR>', opts)
map('n', '<leader>td', '<cmd>Gitsigns toggle_deleted<CR>', opts)
```

Thanks alot for your patience.

**Comment by lewis6991** [comment]
The `on_attach` from lspconfig is **completely unrelated** to the `on_attach` in Gitsigns. The **only** thing they have in common is that they have the same name.

`on_attach` is probaby being run in a `fast event`, in which case you need to do:

```lua
vim.schedule(function()
  print("inside on_attach")
end)
```

**Comment by matu3ba** [comment]
Adding this into the `on_attach = function(bufnr)` function block and starting neovim with `nvim --clean -u minimal.lua ~/dev/git/zig/chepa/src/main.zig` still prints nothing.
Adding `print("outside on_attach")` below the `require('gitsigns').setup{` still works.

**Comment by lewis6991** [comment]
Are the signs showing?

**Comment by matu3ba** [comment]
`nvim --clean -u minimal.lua ~/dev/git/zig/chepa/src/main.zig` shows

![Screenshot_20220331_115330](https://user-images.githubusercontent.com/13089667/161028786-5684adb4-c082-4b2e-aa1c-c4749dd155ca.png)

**Comment by lewis6991** [comment]
Ok, I can reproduce with that, thanks

**Comment by lewis6991** [comment]
Ah sorry, I copied the `minimal.lua` wrong. 

I can't reproduce.

```lua
vim.o.packpath = '/tmp/nvim/site'

local plugins = {
  gitsigns = 'https://github.com/lewis6991/gitsigns.nvim',
}


for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/pack/test/start/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
end

require('gitsigns').setup{
  debug_mode = true, 
  on_attach = function()
    error('DEBUG')
  end
}
```

with `nvim --clean -u minimal.lua <file>`

causes:

```
Error executing vim.schedule lua callback: ...vim/site/pack/test/start/gitsigns/lua/gitsigns/async.lua:68: The coroutine failed
 with this message: [REDACTED]/mininal.lua:20: DEBUG

```

as it should

**Comment by matu3ba** [comment]
lol. I have no idea, but it works now with
```lua
require('gitsigns').setup{
  -- unchanged default settings
  on_attach = function(bufnr)
    local gs = package.loaded.gitsigns

    local function map(mode, l, r, opts)
      opts = opts or {}
      opts.buffer = bufnr
      vim.keymap.set(mode, l, r, opts)
    end

    -- Navigation
    map('n', ']c', function()
      if vim.wo.diff then return ']c' end
      vim.schedule(function() gs.next_hunk() end)
      return '<Ignore>'
    end, {expr=true})

    map('n', '[c', function()
      if vim.wo.diff then return '[c' end
      vim.schedule(function() gs.prev_hunk() end)
      return '<Ignore>'
    end, {expr=true})

    -- Actions
    map({'n', 'v'}, '<leader>hs', ':Gitsigns stage_hunk<CR>')
    map({'n', 'v'}, '<leader>hr', ':Gitsigns reset_hunk<CR>')
    map('n', '<leader>hS', gs.stage_buffer)
    map('n', '<leader>hu', gs.undo_stage_hunk)
    map('n', '<leader>hR', gs.reset_buffer)
    map('n', '<leader>hp', gs.preview_hunk)
    map('n', '<leader>hb', function() gs.blame_line{full=true} end)
    map('n', '<leader>tb', gs.toggle_current_line_blame)
    map('n', '<leader>hd', gs.diffthis)
    map('n', '<leader>hD', function() gs.diffthis('~') end)
    map('n', '<leader>td', gs.toggle_deleted)

    -- Text object
    map({'o', 'x'}, 'ih', ':<C-U>Gitsigns select_hunk<CR>')
  end
}
```

**Comment by dadav** [comment]
Had the same symptoms...`PackerCompile` fixed it for me.
